benchmark/rocblas/utils.hpp

```diff
@@ -0,0 +1,206 @@
+  void copyD2H(const T* dPtr, T* hPtr) {
+    if (hPtr != nullptr) {
+      CHECK_HIP_ERROR(hipMemcpy(hPtr, dPtr, bytes_, hipMemcpyDeviceToHost));
+    }
+    return;
+  }
+
+  // Free device memory
+  void free(T* d) {
+    if (d != nullptr) {
+      CHECK_HIP_ERROR((hipFree)(d));
+    }
+  }
+};
+
+// Pseudo-vector subclass which uses device memory
+template <typename T, bool CopyToHost = false>
+class HIPVector : private HIPDeviceMemory<T> {
+ public:
+  explicit HIPVector(size_t s) : HIPDeviceMemory<T>(s) {
+    d_data_ = this->alloc();
+  }
+
+  // Constructor using host pointer copies data to device
+  HIPVector(size_t s, T* hPtr) : HIPVector<T, CopyToHost>(s) {
+    h_data_ = hPtr;
+    this->copyH2D(h_data_, d_data_);
+  }
+
+  // Destructor copies data back to host if specified & valid
+  // & free-up device memory
+  ~HIPVector() {
+    if constexpr (CopyToHost) {
+      this->copyD2H(d_data_, h_data_);
+    }
+    this->free(d_data_);
+  }
+
+  // Decay into device pointer wherever pointer is expected
+  operator T*() { return d_data_; }
+  operator const T*() const { return d_data_; }
+  T* data() const { return d_data_; }
+
+  // Disallow copying or assigning
+  HIPVector(const HIPVector&) = delete;
+  HIPVector& operator=(const HIPVector&) = delete;
+
+ private:
+  T* d_data_;
+  T* h_data_ = nullptr;
+};
+
+// Pseudo-scalar subclass which uses device memory
+template <typename T, bool CopyToHost = false>
+class HIPScalar : private HIPDeviceMemory<T> {
+ public:
+  explicit HIPScalar() : HIPDeviceMemory<T>() { d_data_ = this->alloc(); }
+
+  // Constructor using host scalar reference copies value to device
+  HIPScalar(T& hValue) : HIPScalar<T, CopyToHost>() {
+    h_data_ = &hValue;
+    this->copyH2D(h_data_, d_data_);
+  }
+
+  // Destructor copies data back to host if specified & valid
+  // & free-up device memory
+  ~HIPScalar() {
+    if constexpr (CopyToHost) {
+      this->copyD2H(d_data_, h_data_);
+    }
+    this->free(d_data_);
+  }
+
+  // Decay into device pointer wherever pointer is expected
+  operator T*() { return d_data_; }
+  operator const T*() const { return d_data_; }
+  T* data() const { return d_data_; }
+
+  // Disallow copying or assigning
+  HIPScalar(const HIPScalar&) = delete;
+  HIPScalar& operator=(const HIPScalar&) = delete;
+
+ private:
+  T* d_data_;
+  T* h_data_ = nullptr;
+};
+
+/**
+ * @fn timef_hip
+ * @brief Calculates the time spent executing the function func returning 2
+ * hipEvents (both overall and HIP events time, returned in nanoseconds in a
+ * tuple of double)
+ */
+template <typename function_t, typename... args_t>
+static inline std::tuple<double, double> timef_hip(function_t func,
+                                                   args_t&&... args) {
+  auto start = std::chrono::system_clock::now();
+  std::vector<hipEvent_t> events = func(std::forward<args_t>(args)...);
+  auto end = std::chrono::system_clock::now();
+
+  double overall_time = (end - start).count();
```

Comment:
Thank you so much for your patch, @OuadiElfarouki!

Are we sure that the `overall_time` will be expressed in nanoseconds in any case? Shall we use `std::chrono::nanoseconds` to be on the safe side?

Reply:
Thanks @pgorlani. Actually doc says that `nanoseconds` is the highest resolution and is thus always used by default. All `overall_time` computations in SYCL-BLAS are assuming this as well.